### ISSUE TITLE
Training in DistributedDataParallel

### ISSUE THREAD

**Issue by Ma-Weijian** [samxuxiang/hnc-cad#3]
Hi, Sam.

Fantastic work on CAD conditional Generation!

I just wonder if i train the model using DistributedDataParallel, will the codebook on each GPU remain the same? I tried to use your update code on multiple GPUs using DistributedDataParallel, but it seems that the number of updated codes on each GPU is not the same when I uncomment the print code.

I know the training data will be shuffled across GPUs during training, but such reinitialization scheme made me doubt whether the codebook will remain the same after several epochs.

Worse still, when I asked New Bing about whether the codebook will remain the same after backward of each batch. New Bing tells me No.....

Now I really have no idea about how the reinitialization scheme and the codebook works on DDP. 

Any idea helps. And, happy Moon Festival!

Weijian.

**Comment by Ma-Weijian** [comment]
Well, it seems that i am sure New Bing is wrong. The gradient on each card remains the same in ddp. So the codebook remains the same as long as it is upgraded by gradients only.

However, it is really intersting to think about what will happen when the codebook on each card is updated independently. The performance will be better, worse or ramains the same compared with syncing all the gradients?

Any idea helps.
